x-pack/plugins/reporting/server/plugin.ts

```diff
@@ -22,61 +20,75 @@ export class ReportingPlugin
   private readonly initializerContext: PluginInitializerContext<ReportingConfigType>;
   private logger: LevelLogger;
   private reportingCore?: ReportingCore;
-  private config$: Observable<ReportingConfigType>;
+
+  // Setup some observables for modules that need to await setup/start
+  public readonly setup$ = new Rx.Subject<boolean>();
+  public readonly start$ = new Rx.Subject<boolean>();
 
   constructor(context: PluginInitializerContext<ReportingConfigType>) {
     this.logger = new LevelLogger(context.logger.get());
     this.initializerContext = context;
-    this.config$ = context.config.create<ReportingConfigType>();
   }
 
-  public async setup(core: CoreSetup, plugins: ReportingSetupDeps) {
+  public setup(core: CoreSetup, plugins: ReportingSetupDeps) {
     const { elasticsearch, http } = core;
     const { licensing, security } = plugins;
     const { initializerContext: initContext } = this;
     const router = http.createRouter();
     const basePath = http.basePath.get;
 
-    const coreConfig = await createConfig$(core, this.config$, this.logger)
-      .pipe(first())
-      .toPromise(); // apply computed defaults to config
-    const reportingConfig = buildConfig(initContext, core, coreConfig); // combine kbnServer configs
-    this.reportingCore = new ReportingCore(reportingConfig);
-
-    const browserDriverFactory = await createBrowserDriverFactory(reportingConfig, this.logger);
-
-    this.reportingCore.pluginSetup({
-      browserDriverFactory,
-      elasticsearch,
-      licensing,
-      basePath,
-      router,
-      security,
-    });
+    // async background setup
+    buildConfig(initContext, core, this.logger).then((config) => {
+      const reportingCore = new ReportingCore(config);
```

Comment:
I wonder what happens if there is a promise rejection somewhere in this callback? I think it would mean `initializeBrowserDriverFactory` waits forever...

Reply:
https://github.com/tsullivan/kibana/pull/7/files